src/ameba/ast/visitors/implicit_return_visitor.cr

```diff
@@ -0,0 +1,268 @@
+    end
+
+    def visit(node : Crystal::EnumDef) : Bool
+      @rule.test(@source, node, @stack > 0)
+
+      node.members.each &.accept(self)
+
+      false
+    end
+
+    def visit(node : Crystal::Assign | Crystal::OpAssign) : Bool
+      @rule.test(@source, node, @stack > 0)
+
+      incr_stack { node.value.accept(self) }
+
+      false
+    end
+
+    def visit(node : Crystal::MultiAssign) : Bool
+      @rule.test(@source, node, @stack > 0)
+
+      node.targets.each &.accept(self)
+      incr_stack { node.values.each &.accept(self) }
+
+      false
+    end
+
+    def visit(node : Crystal::If | Crystal::Unless) : Bool
+      @rule.test(@source, node, @stack > 0)
+
+      incr_stack { node.cond.accept(self) }
+      node.then.accept(self)
+      node.else.accept(self)
+
+      false
+    end
+
+    def visit(node : Crystal::While | Crystal::Until) : Bool
+      @rule.test(@source, node, @stack > 0)
+
+      incr_stack { node.cond.accept(self) }
+      node.body.accept(self)
+
+      false
+    end
+
+    def visit(node : Crystal::Def) : Bool
+      @rule.test(@source, node, @stack > 0)
+
+      incr_stack do
+        node.args.each &.accept(self)
+        node.double_splat.try &.accept(self)
+        node.block_arg.try &.accept(self)
+      end
+
+      if (return_type = node.return_type).is_a?(Crystal::Path)
+        # Special case of the return type being nil, meaning the last
+        # line of the method body is ignored
+        if return_type.names.join("::").in?("::Nil", "Nil")
+          node.body.accept(self)
+        else
+          incr_stack { node.body.accept(self) }
+        end
+      else
+        incr_stack { node.body.accept(self) }
+      end
+
+      false
+    end
+
+    def visit(node : Crystal::Annotation) : Bool
+      @rule.test(@source, node, @stack > 0)
+
+      incr_stack do
+        node.args.each &.accept(self)
+        node.named_args.try &.each &.accept(self)
+      end
+
+      false
+    end
+
+    def visit(node : Crystal::TypeDeclaration) : Bool
+      incr_stack { node.value.try &.accept(self) }
+
+      false
+    end
+
+    def visit(node : Crystal::Macro | Crystal::MacroIf | Crystal::MacroFor) : Bool
+      false
+    end
+
+    def visit(node : Crystal::UninitializedVar) : Bool
+      false
+    end
+
+    def visit(node : Crystal::ArrayLiteral | Crystal::TupleLiteral) : Bool
+      @rule.test(@source, node, @stack > 0)
+
+      incr_stack { node.elements.each &.accept(self) }
+
+      false
+    end
+
+    def visit(node : Crystal::StringInterpolation) : Bool
+      @rule.test(@source, node, @stack > 0)
+
+      node.expressions.each do |exp|
+        unless exp.is_a?(Crystal::StringLiteral)
+          incr_stack { exp.accept(self) }
+        end
+      end
+
+      false
+    end
+
+    def visit(node : Crystal::HashLiteral | Crystal::NamedTupleLiteral) : Bool
+      @rule.test(@source, node, @stack > 0)
+
+      incr_stack { node.entries.each &.value.accept(self) }
+
+      false
+    end
```

Comment:
many of these methods return `false` meaning the nested AST will not expand further. However, what if HashLiteral has nested hashes or similar examples?

Reply:
They return false because I don't want to use the built in accept_children methods (which is what this bool indicates). Line 170 there has it visiting all of the entries of the hash, and because they're used in the hash, increments the stack.

Reply:
But that means we try to manually accept the children instead of delegating this work to the visitor. If a new node would be added to the syntax tree in crystal compiler, we will never visit it without changing this code.

Reply:
Yes, that's true. We can't delegate this to the stdlib visitor though as we need to handle how the child elements are called, as different parts of elements are "used" or not (i.e., the condition of an `if` statement is used by itself, but the body and else are used by its parent).

The stdlib parser is internal code to the Crystal compiler, so using it requires some level of being okay with breaking changes as-is.